### DISCUSSION TITLE
Production build with Parcel

### DISCUSSION THREAD

**Discussion by JennaSys** [floating-ui/floating-ui#1917] [React] [answered]
This may be a Mantine issue, but I'm getting a runtime error in the floating-ui lib with a production build using the Parcel bundler:
```
TypeError: Cannot destructure property 'open' of 'context' as it is undefined.
    at $10d570edfd0c0404$export$ae780daf29e6d456 (floating-ui.react-dom-interactions.esm.js:971:5)
    at $b592912bba5282fb$export$1c4b08e0eca38426 (use-tooltip.ts:38:13)
    at Tooltip.tsx:126:29
    at $833559fe574b4225$var$Xh (react-dom.production.min.js:167:137)
    at $833559fe574b4225$var$Zi (react-dom.production.min.js:193:57)
    at $833559fe574b4225$var$Wk (react-dom.production.min.js:294:275)
    at $833559fe574b4225$var$Vk (react-dom.production.min.js:280:389)
    at $833559fe574b4225$var$Uk (react-dom.production.min.js:280:320)
    at $833559fe574b4225$var$Jk (react-dom.production.min.js:280:180)
    at $833559fe574b4225$var$Ok (react-dom.production.min.js:271:88)
```
It does NOT happen in a debug build though. It looks like the `context` provided by `useFloating()` that is passed in to `useHover()`is not valid. My question is, what could be different with a production build that might cause this issue? I'm out of ideas to fix it so I thought I'd ask here if anyone has any ideas.

<details><summary>Relevant code from Mantine use-tooltip():</summary>

```javascript
  const {
    x,
    y,
    reference,
    floating,
    context,
    refs,
    update,
    placement,
    middlewareData: { arrow: { x: arrowX, y: arrowY } = {} },
  } = useFloating({
    placement: settings.position,
    open: opened,
    onOpenChange: onChange,
    middleware: [
      offset(settings.offset),
      shift({ padding: 8 }),
      flip(),
      arrow({ element: settings.arrowRef }),
      ...(settings.inline ? [inline()] : []),
    ],
  });

  const { getReferenceProps, getFloatingProps } = useInteractions([
    useHover(context, {
      enabled: settings.events.hover,
      delay: withinGroup ? groupDelay : { open: settings.openDelay, close: settings.closeDelay },
      mouseOnly: !settings.events.touch,
    }),
    useFocus(context, { enabled: settings.events.focus, keyboardOnly: true }),
    useRole(context, { role: 'tooltip' }),
    // cannot be used with controlled tooltip, page jumps
    useDismiss(context, { enabled: typeof settings.opened === undefined }),
    useDelayGroup(context, { id: uid }),
  ]);
```
</details>

**Comment by atomiks** [comment]
Hard to say without a sample of the compiled code but it sounds like a minification bug. Have you tried updating all the packages? (Parcel's that is)

**Reply by JennaSys** [reply]
It's giving me the same error with and without minification, but I'll make sure Parcel is latest and try again.

**Reply by JennaSys** [reply]
I was already running the latest version of Parcel. I did figure out that when I disable scope hoisting, the problem goes away  (but my build is then 4x the size).

**Reply by JennaSys** [reply]
Not sure if it's related, but [this Parcel issue](https://github.com/parcel-bundler/parcel/issues/8310) seems to indicate a similar behavior to what I'm experiencing.

**Reply by atomiks** [reply]
I guess it's a bug in Parcel's scope hoisting implementation then. Would be best to file an issue about it with the sample output of where it goes wrong.

**Reply by JennaSys** [reply]
I'll point to this discussion on that issue. If it's related, there may be a code pattern in [useFloating.ts](https://github.com/floating-ui/floating-ui/blob/master/packages/react-dom-interactions/src/useFloating.ts) that is causing Parcel to lose track of a reference rather than bail-out when it should.

**Comment by JennaSys** [comment]
I've confirmed the issue is in [useFloating.ts](https://github.com/floating-ui/floating-ui/blob/08371b9b707f38172848457ad09a1432e22000f1/packages/react-dom-interactions/src/useFloating.ts#L16) and likely has something to do with exporting a name that is the same as one you are importing.

`import {useFloating as usePosition} from '@floating-ui/react-dom';`

`export function useFloating<RT extends ReferenceType = ReferenceType>({`

As a test, renaming the `useFloating()` function to `useFloating2()` and changing the import in the Mantine [usetooltip module](https://github.com/mantinedev/mantine/blob/97d7b75d501b596012fde7b212cd7e1bea16e8e7/src/mantine-core/src/Tooltip/use-tooltip.ts#L3) fixes the issue with Parcel scope hoisting that I am seeing.

I'll post this info to the open Parcel issue, but it might be something to consider on the floating-ui side.

**Comment by JennaSys** [comment, marked as answer]
Just FYI, I posted this as an [issue](https://github.com/parcel-bundler/parcel/issues/8652) with the parcel bundler, and the [v2.81 build](https://github.com/parcel-bundler/parcel/releases/tag/v2.8.1) of Parcel seems to have fixed the issue I was having with scope hoisting.
